ibis/expr/types/arrays.py

```diff
@@ -1108,15 +1116,49 @@ def array(values: Iterable[V]) -> ArrayValue:
     │ [3, 42, ... +1]      │
     └──────────────────────┘
 
-    >>> ibis.array([t.a, 42 + ibis.literal(5)])
+    >>> ibis.array([t.a, 42 + ibis.literal(5)], type="array<float>")
     ┏━━━━━━━━━━━━━━━━━━━━━━┓
     ┃ Array()              ┃
     ┡━━━━━━━━━━━━━━━━━━━━━━┩
-    │ array<int64>         │
+    │ array<float64>       │
     ├──────────────────────┤
-    │ [1, 47]              │
-    │ [2, 47]              │
-    │ [3, 47]              │
+    │ [1.0, 47.0]          │
+    │ [2.0, 47.0]          │
+    │ [3.0, 47.0]          │
     └──────────────────────┘
     """
-    return ops.Array(tuple(values)).to_expr()
+    type = dt.dtype(type) if type is not None else None
+    if type is not None and not isinstance(type, dt.Array):
+        raise ValidationError(f"type must be an array, got {type}")
+
+    if isinstance(values, ir.Value):
+        if type is not None:
+            return values.cast(type)
+        elif isinstance(values, ArrayValue):
+            return values
+        else:
+            raise ValidationError(
+                f"If no type passed, values must be an array, got {values.type()}"
+            )
+
+    if values is None:
+        if type is None:
+            raise ValidationError("If values is None/NULL, type must be provided")
+        return ir.null(type)
```

Comment:
I don't like this. It's unnecessary to have to support another way to construct a null value.

If someone wants a `NULL` with an array type, they should use the `null` function.

Reply:
I have a thought here but I better do it at my computer

Reply:
See these tests in mismo:
- [test_jaccard](https://github.com/NickCrews/mismo/blob/e536163d779f4b0a5b2c6dd13a0d3c4587c6250d/mismo/sets/tests/test_compare.py#L10-L27)
- [test_array_min](https://github.com/NickCrews/mismo/blob/e536163d779f4b0a5b2c6dd13a0d3c4587c6250d/mismo/arrays/tests/test_array.py#L13-L30)
- [test_array_mean_median](https://github.com/NickCrews/mismo/blob/e536163d779f4b0a5b2c6dd13a0d3c4587c6250d/mismo/arrays/tests/test_array.py#L53-L80)

I need to construct an `array<int>` with a variety of values:
- NULL
- []
- [1,2,3]

There are three different ibis APIs for doing this:
- `ibis.null(type="array<int>")`: this only works for the NULL case. I would need some if/else branching in my test setup to use this for the NULL case, but would need something else for the other cases.
- `ibis.literal(val, type="array<int>")`. This is currently what I do, and it works for all 3 inputs here, no if/else branching required. But the docstring for literal() says that it is going to get deprecated soon for constructing complex types, and I should use array()/map()/struct() instead.
- `ibis.array()`: If we keep the functionality the way I propose here, then I can do `ibis.array(val, type="array<int>")` and it works for all three input cases. If we don't allow passing in None, then I would need to do some if/else branching to use ibis.null() in that case.

I think it can be summarized as SOMEONE is gonna need to do this `if values is None:` branching, and I would rather have it be down here in one place in library code, than in many places in user code.